server/schedulers/evict_slow_trend.go

```diff
@@ -0,0 +1,437 @@
+		return
+	}
+
+	store := candidates[0]
+	if !checkStoreSlowerThanOthers(cluster, store) {
+		storeSlowTrendActionStatusGauge.WithLabelValues("cand.none:not-slower").Inc()
+		return
+	}
+	return store
+}
+
+func checkStoresAreUpdated(cluster schedule.Cluster, baseline *core.StoreInfo) bool {
+	stores := cluster.GetStores()
+	if len(stores) <= 3 {
+		return false
+	}
+	expected := (len(stores) + 1) / 2
+	updatedStores := 0
+	baselineTs := baseline.GetLastHeartbeatTS()
+	for _, store := range stores {
+		if store.IsRemoved() {
+			updatedStores += 1
+			continue
+		}
+		if !(store.IsPreparing() || store.IsServing()) {
+			updatedStores += 1
+			continue
+		}
+		if store.GetID() == baseline.GetID() {
+			updatedStores += 1
+			continue
+		}
+		if baselineTs.Before(store.GetLastHeartbeatTS()) {
+			updatedStores += 1
+		}
+	}
+	storeSlowTrendMiscGauge.WithLabelValues("stores.check-updated:count").Set(float64(updatedStores))
+	storeSlowTrendMiscGauge.WithLabelValues("stores.check-updated:expected").Set(float64(expected))
+	return updatedStores >= expected
+}
+
+func checkStoreSlowerThanOthers(cluster schedule.Cluster, target *core.StoreInfo) bool {
+	stores := cluster.GetStores()
+	expected := (len(stores)*2 + 1) / 3
+	if expected >= len(stores) {
+		expected = len(stores) - 1
+	}
+	if expected < 1 {
+		expected = 1
+	}
+	targetSlowTrend := target.GetSlowTrend()
+	if targetSlowTrend == nil {
+		storeSlowTrendActionStatusGauge.WithLabelValues("cand.check-slower:no-data").Inc()
+		return false
+	}
+	slowerThanStoresNum := 0
+	for _, store := range stores {
+		if store.IsRemoved() {
+			continue
+		}
+		if !(store.IsPreparing() || store.IsServing()) {
+			continue
+		}
+		if store.GetID() == target.GetID() {
+			continue
+		}
+		slowTrend := store.GetSlowTrend()
+		// Use `SlowTrend.ResultValue` at first, but not good, `CauseValue` is better
+		// Greater `CuaseValue` means slower
+		if slowTrend != nil && slowTrend.CauseValue < targetSlowTrend.CauseValue && slowTrend.CauseValue > 0 {
+			slowerThanStoresNum += 1
+		}
+	}
+	storeSlowTrendMiscGauge.WithLabelValues("store.check-slower:count").Set(float64(slowerThanStoresNum))
+	storeSlowTrendMiscGauge.WithLabelValues("store.check-slower:expected").Set(float64(expected))
+	return slowerThanStoresNum >= expected
+}
+
+func checkStoreCanRecover(cluster schedule.Cluster, target *core.StoreInfo) bool {
+	slowTrend := target.GetSlowTrend()
+	if slowTrend == nil || slowTrend.CauseRate >= 0 {
+		storeSlowTrendActionStatusGauge.WithLabelValues("recover.reject:no-event").Inc()
+		return false
+	} else {
+		storeSlowTrendActionStatusGauge.WithLabelValues("recover.judging:got-event").Inc()
+	}
+	return checkStoreFasterThanOthers(cluster, target)
+}
+
+func checkStoreFasterThanOthers(cluster schedule.Cluster, target *core.StoreInfo) bool {
+	stores := cluster.GetStores()
+	expected := (len(stores) + 1) / 2
+	if expected >= len(stores) {
+		expected = len(stores) - 1
+	}
+	if expected < 1 {
+		expected = 1
+	}
```

Comment:
ditto

Reply:
It's not the same, the original can handle `len(stores) == 1`

Reply:
It's redundant, removed